### PR TITLE
BUG: groupby reorders categorical categories

### LINKED ISSUE #48749: API/PERF: Don't reorder categoricals when grouping by an unordered categorical and `sort=False`
xref: https://github.com/dask/dask/pull/9486#issue-1372066649

TLDR: When calling `df.groupby(key=categocial<order=False>, sort=True, observed=False)` the resulting `CategoricalIndex` will have it's values _and categories_ unordered. 

```
In [1]:     df = DataFrame(
   ...:         [
   ...:             ["(7.5, 10]", 10, 10],
   ...:             ["(7.5, 10]", 8, 20],
   ...:             ["(2.5, 5]", 5, 30],
   ...:             ["(5, 7.5]", 6, 40],
   ...:             ["(2.5, 5]", 4, 50],
   ...:             ["(0, 2.5]", 1, 60],
   ...:             ["(5, 7.5]", 7, 70],
   ...:         ],
   ...:         columns=["range", "foo", "bar"],
   ...:     )

In [2]: col = "range"

In [3]: df["range"] = Categorical(df["range"], ordered=False)

In [4]: df.groupby(col, sort=True, observed=False).first().index
Out[4]: CategoricalIndex(['(0, 2.5]', '(2.5, 5]', '(5, 7.5]', '(7.5, 10]'], categories=['(0, 2.5]', '(2.5, 5]', '(5, 7.5]', '(7.5, 10]'], ordered=False, dtype='category', name='range')

In [5]: df.groupby(col, sort=False, observed=False).first().index
Out[5]: CategoricalIndex(['(7.5, 10]', '(2.5, 5]', '(5, 7.5]', '(0, 2.5]'], categories=['(7.5, 10]', '(2.5, 5]', '(5, 7.5]', '(0, 2.5]'], ordered=False, dtype='category', name='range')
```

It's reasonable that the values are not sorted, but a lot of extra work can be spent un-ordering the _categories_ in:

https://github.com/pandas-dev/pandas/blob/44a4f1619ff5031e59a970a61fac94c3745e4433/pandas/core/groupby/categorical.py#L77-L92

May have been an outcome of fixing https://github.com/pandas-dev/pandas/issues/8868, but if grouping and `sort=False` the values can be achieved without reordering the categories, there would probably be a nice performance benefit.

### PR DESCRIPTION
**Pull request by rhshadrach**
Ref: #48749

- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

~Still needs more tests and a whatsnew~

### CODE REVIEW

**Comment by rhshadrach** [comment]
@mroeschke - this doesn't tackle the performance aspect of #48749, only the behavior. It's not clear to me but it seems possible there is still something that can be done in regards to performance, so I think we should leave #48749 open for now.

**Comment by mroeschke** [comment]
> I think we should leave https://github.com/pandas-dev/pandas/issues/48749 open for now.

Agreed

Also looks like a ASV benchmark needs addressing: https://github.com/pandas-dev/pandas/actions/runs/3261927947/jobs/5357896059

File: pandas/tests/groupby/test_categorical.py [hunk 1]
```diff
@@ -782,7 +782,7 @@ def test_preserve_categories():
     # ordered=False
     df = DataFrame({"A": Categorical(list("ba"), categories=categories, ordered=False)})
     sort_index = CategoricalIndex(categories, categories, ordered=False, name="A")
-    nosort_index = CategoricalIndex(list("bac"), list("bac"), ordered=False, name="A")
+    nosort_index = CategoricalIndex(list("bac"), list("abc"), ordered=False, name="A")
```
Comment (mroeschke) [inline pandas/tests/groupby/test_categorical.py:785]: Could you add a GH reference by these changed tests?

**Comment by rhshadrach** [comment]
Regarding the failing ASV, not sure how to approach here. I can't reproduce locally with `asv continuous -f 1.1 upstream/main HEAD --quick` and the log doesn't give any indication of what's failing.

**Comment by rhshadrach** [comment]
ASVs:

```
       before           after         ratio
     [890d0975]       [8166c13b]
     <groupby_cat_order~2^2>       <groupby_cat_order>
+         258±4μs      1.31±0.01ms     5.08  groupby.Categories.time_groupby_extra_cat_sort
+       271±0.5μs      1.35±0.01ms     4.98  groupby.Categories.time_groupby_ordered_sort
+       270±0.8μs      1.34±0.01ms     4.97  groupby.Categories.time_groupby_sort
+     1.08±0.02ms      2.34±0.01ms     2.18  groupby.Categories.time_groupby_extra_cat_nosort
+     1.27±0.01ms      2.50±0.01ms     1.97  groupby.Categories.time_groupby_nosort
+     1.44±0.01ms      2.50±0.07ms     1.73  groupby.Categories.time_groupby_ordered_nosort
+         105±5μs          118±2μs     1.12  groupby.GroupByMethods.time_dtype_as_field('int16', 'quantile', 'direct', 1)
-      12.0±0.1ms       10.8±0.3ms     0.90  groupby.AggEngine.time_dataframe_cython(True)
```

These benchmarks use data with length 100000 and 10000 categories.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Awesome, thanks @rhshadrach

**Pull request merged**